hal-x86_64/src/mm/mod.rs

```diff
@@ -0,0 +1,831 @@
+        unsafe { Self::from_pml4_page(vm_offset, phys) }
+    }
+    unsafe fn from_pml4_page(vm_offset: VAddr, page: Page<PAddr, Size4Kb>) -> &'static mut Self {
+        let pml4_paddr = page.base_address();
+        tracing::trace!(?pml4_paddr, ?vm_offset);
+
+        let virt = vm_offset + VAddr::from_usize(pml4_paddr.as_usize());
+        tracing::debug!(current_pml4_vaddr = ?virt);
+        &mut *(virt.as_ptr::<Self>() as *mut _)
+    }
+}
+
+impl<'mapper, A> Map<'mapper, Size4Kb, A> for PageCtrl
+where
+    A: page::Alloc<Size4Kb>,
+{
+    type Handle = Handle<'mapper, level::Pt>;
+    // type Handle =
+    /// Map the virtual memory page represented by `virt` to the physical page
+    /// represented bt `phys`.
+    ///
+    /// # Panics
+    ///
+    /// - If the physical address is invalid.
+    /// - If the page is already mapped.
+    fn map_page(
+        &'mapper mut self,
+        virt: Page<VAddr, Size4Kb>,
+        phys: Page<PAddr, Size4Kb>,
+        frame_alloc: &mut A,
+    ) -> Self::Handle {
+        let span = tracing::debug_span!("map_page", ?virt, ?phys);
+        let _e = span.enter();
+        let pml4 = unsafe { self.pml4.as_mut() };
+
+        let vaddr = virt.base_address();
+        tracing::trace!(?vaddr);
+
+        let page_table = pml4
+            .create_next_table(virt, frame_alloc)
+            .create_next_table(virt, frame_alloc)
+            .create_next_table(virt, frame_alloc);
+        tracing::debug!(?page_table);
+
+        let entry = &mut page_table[virt];
+        assert!(
+            !entry.is_present(),
+            "mapped page table entry already in use"
+        );
+        assert!(!entry.is_huge(), "huge bit should not be set for 4KB entry");
+        let entry = entry.set_phys_page(phys).set_present(true);
+        Handle { entry, page: virt }
+    }
+
+    fn flags_mut(&'mapper mut self, _virt: Page<VAddr, Size4Kb>) -> Self::Handle {
+        unimplemented!()
+    }
+
+    /// Unmap the provided virtual page, returning the physical page it was
+    /// previously mapped to.
+    ///
+    /// This does not deallocate any page frames.
+    ///
+    /// # Panics
+    ///
+    /// - If the virtual page was not mapped.
+    fn unmap(&'mapper mut self, _virt: Page<VAddr, Size4Kb>) -> Page<PAddr, Size4Kb> {
+        unimplemented!()
+    }
+}
+
+impl<S> TranslatePage<S> for PageCtrl
+where
+    S: Size,
+    PageTable<level::Pml4>: TranslatePage<S>,
+{
+    fn translate_page(&self, virt: Page<VAddr, S>) -> TranslateResult<PAddr, S> {
+        unsafe { self.pml4.as_ref() }.translate_page(virt)
+    }
+}
+
+impl PageCtrl {
+    pub fn current() -> Self {
+        let vm_offset = VM_OFFSET.load(Ordering::Acquire);
+        assert_ne!(
+            vm_offset,
+            core::usize::MAX,
+            "`init_paging` must be called before calling `PageTable::current`!"
+        );
+        let vm_offset = VAddr::from_usize(vm_offset);
+        let pml4 = PageTable::current(vm_offset);
```

Comment:
We can probably replace this with a hardcoded value, but let's perhaps do this in a follow-up.